debugger/server/debugger_server.go

```diff
@@ -1,40 +1,139 @@
 package server
 
 import (
+	"context"
+	"encoding/json"
 	"fmt"
 	"io"
 	"net"
 	"os"
+	"os/signal"
+	"syscall"
 
 	"github.com/earthly/earthly/conslogging"
+	"github.com/earthly/earthly/debugger/common"
+
+	"github.com/creack/pty"
+	"github.com/hashicorp/yamux"
+	"golang.org/x/crypto/ssh/terminal"
 )
 
-// DebugServer is provides a console which reverse shells connect to
+type session struct {
+	yaSession *yamux.Session
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	ttyCon     net.Conn
+	resizeConn net.Conn
+
+	server *DebugServer
+}
+
+// DebugServer provides a server that accepts a remote debugging shell
 type DebugServer struct {
-	addr    string
+	session *session
 	console conslogging.ConsoleLogger
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	sigs chan os.Signal
+	addr string
 }
 
-// NewDebugServer creates a new debug server
-func NewDebugServer(console conslogging.ConsoleLogger) *DebugServer {
-	return &DebugServer{
-		addr:    "127.0.0.1:0",
-		console: console,
+func (s *session) sendNewWindowSize(size *pty.Winsize) error {
+	b, err := json.Marshal(size)
+	if err != nil {
+		return err
 	}
+	return common.WriteUint16PrefixedData(s.resizeConn, b)
+}
+
+func (s *session) handle() error {
+	for {
+		stream, err := s.yaSession.Accept()
+		if err != nil {
+			return err
+		}
+
+		buf := make([]byte, 1)
+		stream.Read(buf)
+
+		switch buf[0] {
+		case common.PtyStream:
+			go s.handlePtyStream(stream)
+
+		case common.WinChangeStream:
+			s.resizeConn = stream
+			s.server.sigs <- syscall.SIGWINCH
+		default:
+			return fmt.Errorf("unsupported stream code %v", buf[0])
+		}
+	}
+}
+
+func (s *session) handlePtyStream(conn net.Conn) error {
+	go func() {
+		_, _ = io.Copy(os.Stdout, conn)
```

Comment:
Should we log these errors?

Reply:
probably a good idea, I'll ignore any EOF errors though.